protodroid/src/main/java/id/lukasdylan/grpc/protodroid/internal/service/ProtodroidClientCall.kt

```diff
@@ -52,28 +51,28 @@ internal class ProtodroidClientCall<RequestObject, ResponseObject>(
     }
 
     override fun onUpdateStatusState(status: Status?, header: Metadata?) {
-        state = state.copy(status = status)
+        val statusLevel = if (status?.code == Status.Code.OK)
+            DataState.StatusLevel.OK
+        else
+            DataState.StatusLevel.ERROR
+
+        state = state.copy(
+            statusCode = status?.code?.value(),
+            statusLevel = statusLevel,
+            statusDescription = status?.description,
+            statusName = status?.code?.name,
+            statusErrorCause = status?.cause?.message
+        )
         if (state.responseHeader == null) {
-            state = state.copy(responseHeader = header)
+            state = state.copy(responseHeader = header.toString())
         }
     }
 
     override fun onFinalState() {
-        GlobalScope.launch(Dispatchers.IO) {
-            val finalState = state
-            finalState.printLogFullResponse()
-
-            val id = repository.saveNewData(finalState)
-            if (id != -1L) {
-                notificationListener.sendNotification(
-                    title = finalState.serviceName.split("/").getOrElse(1) {
-                        finalState.serviceName
-                    },
-                    message = "${finalState.status?.code?.name} (${finalState.status?.code?.value()})",
-                    dataId = id,
-                    serviceName = finalState.serviceName,
-                )
-            }
+        val finalState = state
+        if (protodroid.loggingEnabled) finalState.printLogFullResponse()
+        launch {
```

Comment:
I think better to use GlobalScope since it not coupled with any application lifecycle, and also not holding Coroutines thing

Reply:
Nice, done